### PR TITLE
Re-add facebook support

### PR DESCRIPTION
**Pull request by manu-chroma**
Addresses https://github.com/manu-chroma/username-availability-checker/issues/20

### CODE REVIEW

File: tests/test_username_api.py [hunk 1]
```diff
@@ -259,63 +267,63 @@ def test_available_twitter_username(self, user):
 
   def test_github_format_checking(self):
     resp = self.app.get('/check/github/{}'.format(invalid_username))
-    json_resp = json.loads(resp.get_data())
+    json_resp = json.loads(resp.get_data().decode())
```
Comment (manu-chroma) [inline tests/test_username_api.py]: I don't know how tests were passing on travis in the first place. ``resp.get_data()`` returns bytes and had to be decoded. @andrewda any idea?

Comment (andrewda) [reply tests/test_username_api.py]: Hmm it was working fine locally for me... Not sure

File: tests/test_username_api.py [hunk 2]
```diff
@@ -73,6 +76,11 @@ def test_taken_instagram_username(self, user):
   def test_taken_twitter_username(self, user):
     assert_response(self.app, 'twitter', user, 200)
 
+  @pytest.mark.parametrize('user', data['facebook']['taken_usernames'])
+  def test_taken_github_username(self, user):
```
Comment (andrewda) [inline tests/test_username_api.py]: Copy/paste issue – should be `test_taken_facebook_username`

Comment (jayvdb) [reply tests/test_username_api.py]: My bad.  I should have spotted this.   All the more reason for https://github.com/manu-chroma/username-availability-checker/issues/36

File: tests/test_username_api.py (hunk 1 above)
Comment (jayvdb) [reply tests/test_username_api.py]: This isnt a Python 2 vs 3 problem?

Is this app Python 2 compatible?  Anyway, maybe someone wants to investigate https://github.com/manu-chroma/username-availability-checker/issues/38 ;-)

File: username_api.py [hunk 3]
```diff
@@ -46,7 +70,11 @@ def check_username(website, username):
 def check_format(website, username):
 	"""Check the format of a username depending on the website"""
 
-	website_parts = patterns['username_patterns'][website]
+	website_parts = patterns.get('username_patterns').get(website, {})
```
Comment (jayvdb) [inline username_api.py]: err, we should always have username patterns for the website ...

here are the facebook published rules : https://www.facebook.com/help/105399436216001#What-are-the-guidelines-around-creating-a-custom-username?

- min: 5 characters (excluding any .)
- permitted characters: A–Z, 0–9, `.`
- not permitted: `.(com|net)`

This is an interesting one, because `.` is permitted and is ignored.  We need a extra pattern in our data for stripping some characters, so that `.` can be stripped from the user input, otherwise the min 5 characters wont be correctly applied.

I am quite sure that this extra pattern will be used by some other websites also.

Facebook also has a fairly common rule that capitalisation can't be used to differentiate usernames.  I think we can assume case-insensitivity as the default for all website, and maybe we need a `case-sensitivity: true` override if we find a website that breaks this rule, and it would be useful to show that the username is available but there is an extremely similar username with a different capitalisation. (Im now on a missing to find a case-sensitive username website ...)

Comment (jayvdb) [reply username_api.py]: Worth noting that there are two parts to case-sensitivity.  One is whether any capitalisation of the username is usable and refers to the same identity.  The other is whether the system allows two usernames with alternative capitalisation.  Wikimedia (MediaWiki is an example of case sensitive usernames , but multiple accounts differing by capitalisation are not permitted)

File: tests/test_username_api.py (hunk 1 above)
Comment (manu-chroma) [reply tests/test_username_api.py]: The problem is both these builds: [#6](https://travis-ci.org/manu-chroma/username-availability-checker/builds/318026284) and [#8](https://travis-ci.org/manu-chroma/username-availability-checker/builds/318231491) use python 3.6 and both of them pass.

File: tests/test_username_api.py [hunk 4]
```diff
@@ -257,66 +261,108 @@ def test_available_twitter_username(self, user):
...
     } == json_resp
+
+  def test_facebook_format_checking(self):
+    invalid_username_facebook = ['hello..world', 'hell...', 'manvendra.com']
+
+    for invalid_username in invalid_username_facebook:
+      resp = self.app.get('/check/facebook/{}'.format(invalid_username))
+      json_resp = json.loads(resp.get_data().decode())
+      assert {
+      'possible': False,
```
Comment (jayvdb) [inline tests/test_username_api.py]: indent?

File: tests/test_username_api.py [hunk 5]
```diff
@@ -257,66 +261,108 @@ def test_available_twitter_username(self, user):
...
+    json_resp = json.loads(resp.get_data().decode())
     assert {
       'possible': False,
       'url': 'https://in.pinterest.com/{}/'.format(invalid_username)
     } == json_resp
 
   def test_instagram_format_checking(self):
     resp = self.app.get('/check/instagram/{}'.format(invalid_username))
-    json_resp = json.loads(resp.get_data())
+    json_resp = json.loads(resp.get_data().decode())
```
Comment (jayvdb) [inline tests/test_username_api.py]: move these to a separate commit?

File: username_api.py (hunk 3 above)
Comment (jayvdb) [reply username_api.py]: This change isnt needed any more.  If you merge this change, other people wont be forced to add this very important metadata for other websites.

File: username_api.py [hunk 6]
```diff
@@ -46,7 +70,11 @@ def check_username(website, username):
...
+
+	# don't check if we do not have RegEx data for a website
```
Comment (jayvdb) [inline username_api.py]: but we should

File: username_api.py [hunk 7]
```diff
@@ -36,6 +36,30 @@ def check_username(website, username):
...
+		res = r.get(url)
+		code = res.status_code
+		# Using mfacebook for checking username,
+		# when a username exists but hidden from
+		# search engines, it gives a login redirect
+		# and 200 code but in case of no profile
+		# available, gives a 404 error.
+
+		# Don't pass url for hidden profiles,
+		# since they will not be accessible
```
Comment (manu-chroma) [inline username_api.py]: this comment is redundant. will remove it

File: websites.yml [hunk 8]
```diff
@@ -38,3 +39,10 @@ username_patterns:
     characters: a-zA-Z0-9_-
     min_length: 1
     max_length: 25
+  facebook:
+    characters: a-z0-9A-Z.
+    min_length: 5
+    max_length: 50
+    invalid_patterns:
+      - "\\.(com|net)"
+      - "(\\.)\\1{1,}" # consecutive '.' not allowed
```
Comment (jayvdb) [inline websites.yml]: two spaces before `#`

File: tests/test_username_api.py (hunk 5 above)
Comment (manu-chroma) [reply tests/test_username_api.py]: done
